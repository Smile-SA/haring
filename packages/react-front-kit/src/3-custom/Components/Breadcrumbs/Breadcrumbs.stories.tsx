import type { Meta, StoryObj } from '@storybook/react';

import { expect } from '@storybook/jest';
import { within } from '@storybook/testing-library';

import { Breadcrumbs as Cmp } from './Breadcrumbs';

const meta = {
  component: Cmp,
  tags: ['autodocs'],
  title: '3-custom/Components/Breadcrumbs',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const Breadcrumbs: IStory = {
  args: {
    children: [
      <a key={1} data-testid="Breadcrumbs-first-element" href="test">
        test
      </a>,
      <a key={2} data-testid="Breadcrumbs-second-element" href="test">
        test
      </a>,
      <a key={3} data-testid="Breadcrumbs-last-element" href="test">
        test
      </a>,
    ],
    separator: '>',
    separatorStyle: {
      color: 'grey',
      fontSize: '16px',
    },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await expect(canvas.getByTestId('Breadcrumbs')).toBeVisible();
    await expect(
      canvas.getByTestId('Breadcrumbs-second-element')
    ).toBeVisible();
    await expect(canvas.getByTestId('Breadcrumbs-last-element')).toBeVisible();
    await expect(await (await canvas.findAllByText('>')).length).toEqual(2);
  },
};
