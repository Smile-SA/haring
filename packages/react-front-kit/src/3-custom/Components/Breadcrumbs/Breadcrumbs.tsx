'use client';

import type { BreadcrumbsProps } from '@mantine/core';
import type { ReactNode } from 'react';

import { Breadcrumbs as MantineBreadcrumbs, createStyles } from '@mantine/core';

interface IBreadcrumbsProps extends BreadcrumbsProps {
  children: ReactNode[];
  separator?: string;
  separatorStyle?: object;
}

export function Breadcrumbs(props: IBreadcrumbsProps): JSX.Element {
  const { separator, separatorStyle, children } = props;

  const useStyles = createStyles((theme) => ({
    breadcrumb: {
      '&:first-of-type': {
        fontSize: '16px',
        fontWeight: 600,
      },
      color: theme.colors.dark[6],
      fontSize: '14px',
      fontWeight: 400,
      marginLeft: '1em',
      marginRight: '1em',
      textDecoration: 'none',
      textTransform: 'capitalize',
    },
    separator: {
      ...separatorStyle,
    },
  }));

  const { classes } = useStyles();

  const svgSeparator = (
    <svg
      fill="none"
      height="14"
      viewBox="0 0 14 14"
      width="14"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.25 10.5L8.75 7L5.25 3.5"
        stroke="#0B7285"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  );

  return (
    <MantineBreadcrumbs
      classNames={{
        breadcrumb: classes.breadcrumb,
        separator: classes.separator,
      }}
      data-testid="Breadcrumbs"
      separator={separator ? separator : svgSeparator}
    >
      {children}
    </MantineBreadcrumbs>
  );
}
