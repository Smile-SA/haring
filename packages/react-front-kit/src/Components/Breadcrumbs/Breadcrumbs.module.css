.breadcrumb {
  &:first-of-type {
    font-size: 16px;
    font-weight: 600;
  }

  color: var(--mantine-color-dark-6);
  font-size: 14px;
  font-weight: 400;
  margin-left: 1em;
  margin-right: 1em;
  text-decoration: none;
  text-transform: capitalize;
  /*noinspection CssInvalidAtRule*/
  @mixin smaller-than $mantine-breakpoint-md {
    margin-left: 0;
    margin-right: 0;
  }
}

.container {
  flex-wrap: wrap;
  row-gap: 5px;
}
