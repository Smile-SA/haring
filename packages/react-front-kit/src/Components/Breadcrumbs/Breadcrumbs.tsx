'use client';

import type { BreadcrumbsProps } from '@mantine/core';
import type { ReactElement, ReactNode } from 'react';

import { Breadcrumbs as MantineBreadcrumbs } from '@mantine/core';

import classes from './Breadcrumbs.module.css';

export interface IBreadcrumbsProps extends BreadcrumbsProps {
  children: ReactNode;
  separator?: string;
}

const svgSeparator = (
  <svg
    fill="none"
    height="14"
    viewBox="0 0 14 14"
    width="14"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5.25 10.5L8.75 7L5.25 3.5"
      stroke="#0B7285"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
  </svg>
);

/** Additional props will be forwarded to the [Mantine Breadcrumb component](https://mantine.dev/core/breadcrumbs) */
export function Breadcrumbs(props: IBreadcrumbsProps): ReactElement {
  const { separator = svgSeparator, children, ...BreadcrumbsProps } = props;

  return (
    <MantineBreadcrumbs
      className={classes.container}
      classNames={{
        breadcrumb: classes.breadcrumb,
      }}
      data-testid="Breadcrumbs"
      separator={separator}
      {...BreadcrumbsProps}
    >
      {children}
    </MantineBreadcrumbs>
  );
}
