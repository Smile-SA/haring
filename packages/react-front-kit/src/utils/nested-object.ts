export type INestedObjectWithDepth<O extends object> = O & {
  children?: INestedObjectWithDepth<O>[];
  depth?: number;
};

export function flattenNestedObjects<O extends object>(
  arr: INestedObjectWithDepth<O>[],
): INestedObjectWithDepth<O>[] {
  let children: INestedObjectWithDepth<O>[] = [];
  const flattened = arr.map((m) => {
    if (m.children && m.children.length) {
      children = [...children, ...m.children];
    }
    return m;
  });
  return flattened.concat(
    children.length ? flattenNestedObjects(children) : children,
  );
}

export function assignDepth<O extends object>(
  arr?: INestedObjectWithDepth<O>[],
  depth = 0,
): INestedObjectWithDepth<O>[] {
  (arr ?? []).forEach((o) => {
    o.depth = depth;
    assignDepth(o.children, depth + 1);
  });
  return arr ?? [];
}
