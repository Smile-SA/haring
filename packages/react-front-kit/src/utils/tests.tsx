import type {
  Queries,
  RenderOptions,
  RenderResult,
  queries,
} from '@testing-library/react';
import type { ReactElement, ReactNode } from 'react';

import { render } from '@testing-library/react';

import { Provider } from '../3-custom/Provider/Provider';

export function renderWithProviders<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  ui: ReactElement,
  renderOptions: RenderOptions<Q, Container, BaseElement> = {}
): RenderResult<Q, Container, BaseElement> {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return <Provider>{children}</Provider>;
  }
  return render<Q, Container, BaseElement>(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}
