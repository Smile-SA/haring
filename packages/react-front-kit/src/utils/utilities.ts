export function isNotNullNorEmpty<S>(
  value: Record<string, never> | S,
): value is Exclude<S, null | undefined> {
  return (
    value !== null &&
    value !== undefined &&
    (typeof value === 'object' ? Object.keys(value).length !== 0 : true)
  );
}

export function isCallback<T, U>(maybeFunc: T | U): maybeFunc is T {
  return typeof maybeFunc === 'function';
}
