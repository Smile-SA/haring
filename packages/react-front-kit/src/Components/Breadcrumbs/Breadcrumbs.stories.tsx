import type { Meta, StoryObj } from '@storybook/react';

import { Breadcrumbs as Cmp } from './Breadcrumbs';

const meta = {
  argTypes: {
    children: {
      control: false,
    },
  },
  component: Cmp,
  tags: ['autodocs'],
  title: '3-custom/Components/Breadcrumbs',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const Breadcrumbs: IStory = {
  args: {
    children: [
      <a key={1} data-testid="Breadcrumbs-first-element" href="test">
        test
      </a>,
      <a key={2} data-testid="Breadcrumbs-second-element" href="test">
        test
      </a>,
      <a key={3} data-testid="Breadcrumbs-last-element" href="test">
        test
      </a>,
    ],
  },
};
