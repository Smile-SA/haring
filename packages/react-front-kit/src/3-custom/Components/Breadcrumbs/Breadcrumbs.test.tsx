import { renderWithProviders } from '../../../utils/tests';

// eslint-disable-next-line import/named
import { Breadcrumbs } from './Breadcrumbs';

describe('Breadcrumbs', () => {
  it('matches snapshot', () => {
    const { container } = renderWithProviders(
      <Breadcrumbs
        // eslint-disable-next-line react/no-children-prop
        children={[
          <a key={1} href="test">
            test
          </a>,
          <a key={2} href="test">
            test
          </a>,
        ]}
        separator="/"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
