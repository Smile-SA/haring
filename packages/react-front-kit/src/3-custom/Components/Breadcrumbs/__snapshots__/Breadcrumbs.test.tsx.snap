// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs matches snapshot 1`] = `
<div>
  <div
    class="mantine-Breadcrumbs-root mantine-1ujbd2v"
    data-testid="Breadcrumbs"
  >
    <a
      class="mantine-Breadcrumbs-breadcrumb mantine-eg71k5"
      href="test"
    >
      test
    </a>
    <div
      class="mantine-Text-root mantine-Breadcrumbs-separator mantine-1iubh7d"
    >
      /
    </div>
    <a
      class="mantine-Breadcrumbs-breadcrumb mantine-eg71k5"
      href="test"
    >
      test
    </a>
  </div>
</div>
`;
